File: sdks/python/common/harness_common/environment.py
"""Deployment environment identity, optional access host and observed evidence."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import ClassVar

from harness_common.host import Host


@dataclass(frozen=True, slots=True)
class Environment:
    """A named environment in which Components have running Service instances.

    Host is optional. When omitted, operations use the current execution host.
    It does not identify a Kubernetes node or dictate where the case runner runs.
    """

    name: str
    host: Host | None = field(default=None, kw_only=True)
    kind: ClassVar[str] = "generic"


@dataclass(frozen=True, slots=True)
class HostEnvironment(Environment):
    """A target running directly on the environment's host."""

    kind: ClassVar[str] = "host"


@dataclass
class EnvironmentFacts:
    """Observed facts with provenance. Missing values mean unknown, not false."""

    source: str = ""
    observed_at: str = ""
    values: dict[str, str] = field(default_factory=dict)


@dataclass
class EnvironmentSnapshot:
    """Run evidence; access configuration and credentials are deliberately absent.

    Runner and target facts are separate: a local runner can test a remote Pod.
    Profiles select project-owned conditions and never substitute for observation.
    """

    name: str
    kind: str
    profile: str = "default"
    revision: str = ""
    host_name: str = ""
    host: EnvironmentFacts | None = None
    runner: EnvironmentFacts = field(default_factory=EnvironmentFacts)
    target: EnvironmentFacts = field(default_factory=EnvironmentFacts)

File: sdks/go/common/environment.go
// Package common owns neutral identities and evidence shared by harnesses.
package common

import (
	"fmt"
)

// Environment identifies a deployment environment independently of the runner.
// Host is optional; nil means operations execute on the current host.
// For Kubernetes, Kubeconfig is interpreted on Host, not on the case runner.
type Environment struct {
	Name       string `json:"name" yaml:"name"`
	Kind       string `json:"kind,omitempty" yaml:"kind,omitempty"`
	Host       *Host  `json:"host,omitempty" yaml:"host,omitempty"`
	Kubeconfig string `json:"kubeconfig,omitempty" yaml:"kubeconfig,omitempty"`
	Context    string `json:"context,omitempty" yaml:"context,omitempty"`
}

func (e Environment) ResolvedKind() string {
	if e.Kind != "" {
		return e.Kind
	}
	if e.Kubeconfig != "" || e.Context != "" {
		return "kubernetes"
	}
	return "generic"
}

func (e Environment) Validate() error {
	if e.Host != nil {
		if err := e.Host.Validate(); err != nil {
			return err
		}
	}
	switch e.ResolvedKind() {
	case "generic", "host":
		if e.Kubeconfig != "" || e.Context != "" {
			return fmt.Errorf("%s environment contains Kubernetes access fields", e.ResolvedKind())
		}
	case "kubernetes":
	default:
		return fmt.Errorf("unknown environment kind %q", e.Kind)
	}
	return nil
}

// EnvironmentFacts are observations. An absent key means unknown, not false.
type EnvironmentFacts struct {
	Source     string            `json:"source"`
	ObservedAt string            `json:"observed_at"`
	Values     map[string]string `json:"values"`
}

// EnvironmentSnapshot records realized conditions without connection credentials.
// Project fixtures populate Target from the target, never from the local runner.
type EnvironmentSnapshot struct {
	Name     string            `json:"name"`
	Kind     string            `json:"kind"`
	Profile  string            `json:"profile"`
	Revision string            `json:"revision"`
	HostName string            `json:"host_name,omitempty"`
	Host     *EnvironmentFacts `json:"host,omitempty"`
	Runner   EnvironmentFacts  `json:"runner"`
	Target   EnvironmentFacts  `json:"target"`
}
